File: src/main/java/space/mori/regenblock/RegenBlockQueue.java
package space.mori.regenblock;

import java.util.*;

import org.bukkit.block.data.BlockData;
import org.bukkit.event.*;
import space.mori.regenblock.Event.*;
import org.bukkit.*;
import org.bukkit.entity.*;

public class RegenBlockQueue implements Runnable
{
    private RegenBlock plugin;
    private long queueDelta;
    private long nextConfigSave;
    private long configSaveDelta;
    private HashMap<String, Long> alarmWentOffAt;
    private Vector<RegenBlockBlock> blockQueue;
    
    public RegenBlockQueue(final RegenBlock plugin) {
        this.queueDelta = 1000L;
        this.nextConfigSave = System.currentTimeMillis();
        this.configSaveDelta = 60000L;
        this.alarmWentOffAt = new HashMap<String, Long>();
        this.blockQueue = new Vector<RegenBlockBlock>();
        this.plugin = plugin;
    }
    
    @Override
    public void run() {
        while (true) {
            try {
                Thread.sleep(this.queueDelta);
            }
            catch (InterruptedException e) {
                e.printStackTrace();
            }
            this.checkQueue();
        }
    }
    
    public void addBlock(final RegenBlockBlock block) {
        this.blockQueue.add(block);
    }
    
    public void regenRegion(final String regionName) {
        final long time = System.currentTimeMillis();
        final Vector<RegenBlockBlock> cloneBlockQueue = (Vector<RegenBlockBlock>)this.blockQueue.clone();
        for (final RegenBlockBlock b : cloneBlockQueue) {
            if (b.getRegionName().equalsIgnoreCase(regionName)) {
                b.setRespawnTime(time);
                this.plugin.getConfiguration().setBlock(b.getLocation(), b.getType(), b.getData(), regionName, time);
            }
        }
    }
    
    public void updateRegionRespawnTime(final String regionName, final long respawnTime) {
        final Vector<RegenBlockBlock> cloneBlockQueue = (Vector<RegenBlockBlock>)this.blockQueue.clone();
        for (final RegenBlockBlock rb : cloneBlockQueue) {
            if (rb.getRegionName().equalsIgnoreCase(regionName)) {
                rb.setRespawnTime(respawnTime);
                this.plugin.getConfiguration().setBlock(rb.getLocation(), rb.getType(), rb.getData(), regionName, respawnTime);
            }
        }
    }
    
    public long shiftRegionRespawnTime(final String regionName, final long respawnTime) {
        final Vector<RegenBlockBlock> cloneBlockQueue = (Vector<RegenBlockBlock>)this.blockQueue.clone();
        final long time = System.currentTimeMillis();
        long earliestRespawn = 99999999999999L;
        long latestRespawn = 0L;
        for (final RegenBlockBlock rb : cloneBlockQueue) {
            if (rb.getRegionName().equalsIgnoreCase(regionName) && rb.getRespawnTime() > time) {
                if (earliestRespawn > rb.getRespawnTime()) {
                    earliestRespawn = rb.getRespawnTime();
                }
                if (latestRespawn >= rb.getRespawnTime()) {
                    continue;
                }
                latestRespawn = rb.getRespawnTime();
            }
        }
        if (earliestRespawn == 99999999999999L) {
            return 0L;
        }
        final long dt = respawnTime - earliestRespawn;
        for (final RegenBlockBlock rb2 : cloneBlockQueue) {
            if (rb2.getRegionName().equalsIgnoreCase(regionName) && rb2.getRespawnTime() > time) {
                rb2.setRespawnTime(rb2.getRespawnTime() + dt);
                this.plugin.getConfiguration().setBlock(rb2.getLocation(), rb2.getType(), rb2.getData(), regionName, rb2.getRespawnTime());
            }
        }
        return latestRespawn + dt + dt;
    }
    
    public Long getRegionRespawnTime(final String regionName) {
        final Vector<RegenBlockBlock> cloneBlockQueue = (Vector<RegenBlockBlock>)this.blockQueue.clone();
        final long time = System.currentTimeMillis();
        for (final RegenBlockBlock rb : cloneBlockQueue) {
            if (rb.getRegionName().equalsIgnoreCase(regionName) && rb.getRespawnTime() > time) {
                return rb.getRespawnTime();
            }
        }
        return null;
    }
    
    public void checkQueue() {
        final ArrayList<RegenBlockBlock> respawnedBlocks = new ArrayList<RegenBlockBlock>();
        final long time = System.currentTimeMillis();
        final Vector<RegenBlockBlock> cloneBlockQueue = (Vector<RegenBlockBlock>)this.blockQueue.clone();
        for (final RegenBlockBlock block : cloneBlockQueue) {
            final long respawnTime = block.getRespawnTime();
            final int x = block.getLocation().getBlockX();
            final int y = block.getLocation().getBlockY();
            final int z = block.getLocation().getBlockZ();
            final Material type = block.getType();
            final BlockData data = block.getData();
            final String regionName = block.getRegionName();
            final int sync = block.getSync();
            final int alarmTime = block.getAlarmTime();
            final String worldName = block.getLocation().getWorld().getName();
            final int regionType = block.getRegionType();
            if (sync == 1 && alarmTime != 0 && respawnTime - alarmTime * 1000 < time) {
                if (!this.alarmWentOffAt.containsKey(regionName)) {
                    this.alarmWentOffAt.put(regionName, time);
                    Bukkit.getServer().getPluginManager().callEvent((Event)new RegenBlockEventAlarm(block.getAlarmRadius(), block.getAlarmMessage(), worldName, x, y, z));
                }
                else if (this.alarmWentOffAt.get(regionName) + block.getAlarmTime() * 1000 + 10000L < time) {
                    this.alarmWentOffAt.remove(regionName);
                }
            }
            if (this.readyForRegen(worldName, regionName, respawnTime, x, y, z, type, data, regionType)) {
                respawnedBlocks.add(block);
            }
        }
        if (respawnedBlocks.size() > 0) {
            for (final RegenBlockBlock block : respawnedBlocks) {
                this.blockQueue.remove(block);
            }
            Bukkit.getServer().getPluginManager().callEvent((Event)new RegenBlockEventRespawn(respawnedBlocks));
        }
        if (this.nextConfigSave < time) {
            this.nextConfigSave = time + this.configSaveDelta;
            Bukkit.getServer().getPluginManager().callEvent((Event)new RegenBlockEventConfigSave());
        }
    }
    
    private boolean readyForRegen(final String worldName, final String regionName, final long respawnTime, final int x, final int y, final int z, final Material type, final BlockData data, final int regionType) {
        final long time = System.currentTimeMillis();
        if (time - respawnTime > 86400000L) {
            return true;
        }
        if (respawnTime < time && regionType == 0) {
            return (type) != Material.SAND && type != Material.GRAVEL || this.plugin.getServer().getWorld(worldName).getBlockAt(x, y - 1, z).getType() != Material.AIR;
        }
        return respawnTime < time && regionType == 1 && this.plugin.getServer().getWorld(worldName).getBlockAt(x, y - 1, z).getType() != Material.AIR;
    }
    
    public void broadcastMessage(final int regionAlarmRadius, final String regionAlarmMessage, final String worldName, final int x, final int y, final int z) {
        final Location location = new Location(this.plugin.getServer().getWorld(worldName), (double)x, (double)y, (double)z);
        Player[] onlinePlayers;
        for (int length = (onlinePlayers = this.plugin.getServer().getOnlinePlayers().toArray(new Player[0])).length, i = 0; i < length; ++i) {
            final Player player = onlinePlayers[i];
            if (player.getWorld().getName().equalsIgnoreCase(worldName) && player.getLocation().distance(location) <= regionAlarmRadius) {
                this.plugin.getLog().sendPlayerWarn(player, regionAlarmMessage);
            }
        }
    }
}

File: src/main/kotlin/space/mori/regenblock/RegenBlockQueue.kt
package space.mori.regenblock

import java.util.*

import org.bukkit.block.data.BlockData
import space.mori.regenblock.event.*
import org.bukkit.*
import org.bukkit.entity.*

class RegenBlockQueue(private val plugin: RegenBlock) : Runnable {
    private val queueDelta: Long
    private var nextConfigSave: Long = 0
    private val configSaveDelta: Long
    private val alarmWentOffAt: HashMap<String, Long>
    private val blockQueue: Vector<RegenBlockBlock>

    init {
        this.queueDelta = 1000L
        this.nextConfigSave = System.currentTimeMillis()
        this.configSaveDelta = 60000L
        this.alarmWentOffAt = HashMap()
        this.blockQueue = Vector()
    }

    override fun run() {
        while (true) {
            try {
                Thread.sleep(this.queueDelta)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }

            this.checkQueue()
        }
    }

    fun addBlock(block: RegenBlockBlock) {
        this.blockQueue.add(block)
    }

    fun regenRegion(regionName: String) {
        val time = System.currentTimeMillis()
        val cloneBlockQueue = this.blockQueue.clone() as Vector<RegenBlockBlock>
        for (b in cloneBlockQueue) {
            if (b.regionName!!.equals(regionName, ignoreCase = true)) {
                b.respawnTime = time
                this.plugin.configuration!!.setBlock(b.location!!, b.type!!, b.data!!, regionName, time)
            }
        }
    }

    fun updateRegionRespawnTime(regionName: String, respawnTime: Long) {
        val cloneBlockQueue = this.blockQueue.clone() as Vector<RegenBlockBlock>
        for (rb in cloneBlockQueue) {
            if (rb.regionName!!.equals(regionName, ignoreCase = true)) {
                rb.respawnTime = respawnTime
                this.plugin.configuration!!.setBlock(rb.location!!, rb.type!!, rb.data!!, regionName, respawnTime)
            }
        }
    }

    fun shiftRegionRespawnTime(regionName: String, respawnTime: Long): Long {
        val cloneBlockQueue = this.blockQueue.clone() as Vector<RegenBlockBlock>
        val time = System.currentTimeMillis()
        var earliestRespawn = 99999999999999L
        var latestRespawn = 0L
        for (rb in cloneBlockQueue) {
            if (rb.regionName!!.equals(regionName, ignoreCase = true) && rb.respawnTime > time) {
                if (earliestRespawn > rb.respawnTime) {
                    earliestRespawn = rb.respawnTime
                }
                if (latestRespawn >= rb.respawnTime) {
                    continue
                }
                latestRespawn = rb.respawnTime
            }
        }
        if (earliestRespawn == 99999999999999L) {
            return 0L
        }
        val dt = respawnTime - earliestRespawn
        for (rb2 in cloneBlockQueue) {
            if (rb2.regionName!!.equals(regionName, ignoreCase = true) && rb2.respawnTime > time) {
                rb2.respawnTime = rb2.respawnTime + dt
                this.plugin.configuration!!.setBlock(rb2.location!!, rb2.type!!, rb2.data!!, regionName, rb2.respawnTime)
            }
        }
        return latestRespawn + dt + dt
    }

    fun getRegionRespawnTime(regionName: String): Long? {
        val cloneBlockQueue = this.blockQueue.clone() as Vector<RegenBlockBlock>
        val time = System.currentTimeMillis()
        for (rb in cloneBlockQueue) {
            if (rb.regionName!!.equals(regionName, ignoreCase = true) && rb.respawnTime > time) {
                return rb.respawnTime
            }
        }
        return null
    }

    fun checkQueue() {
        val respawnedBlocks = ArrayList<RegenBlockBlock>()
        val time = System.currentTimeMillis()
        val cloneBlockQueue = this.blockQueue.clone() as Vector<RegenBlockBlock>
        for (block in cloneBlockQueue) {
            val respawnTime = block.respawnTime
            val x = block.location!!.blockX
            val y = block.location!!.blockY
            val z = block.location!!.blockZ
            val type = block.type
            val data = block.data
            val regionName = block.regionName
            val sync = block.sync
            val alarmTime = block.alarmTime
            val worldName = block.location!!.world.name
            val regionType = block.regionType
            if (sync == 1 && alarmTime != 0 && respawnTime - alarmTime * 1000 < time) {
                if (!this.alarmWentOffAt.containsKey(regionName)) {
                    this.alarmWentOffAt[regionName!!] = time
                    Bukkit.getScheduler().runTask(plugin) {
                        Bukkit.getPluginManager()
                            .callEvent(RegenBlockEventAlarm(block.alarmRadius, block.alarmMessage!!, worldName, x, y, z))
                    }
                } else if (this.alarmWentOffAt[regionName]!! + (block.alarmTime * 1000).toLong() + 10000L < time) {
                    this.alarmWentOffAt.remove(regionName)
                }
            }
            if (this.readyForRegen(worldName, regionName, respawnTime, x, y, z, type, data, regionType)) {
                respawnedBlocks.add(block)
            }
        }
        if (respawnedBlocks.size > 0) {
            for (block in respawnedBlocks) {
                this.blockQueue.remove(block)
            }
            Bukkit.getScheduler()
                .runTask(plugin) { Bukkit.getPluginManager().callEvent(RegenBlockEventRespawn(respawnedBlocks)) }
        }
        if (this.nextConfigSave < time) {
            this.nextConfigSave = time + this.configSaveDelta
            Bukkit.getScheduler().runTask(plugin) { Bukkit.getPluginManager().callEvent(RegenBlockEventConfigSave()) }
        }
    }

    private fun readyForRegen(
        worldName: String,
        regionName: String?,
        respawnTime: Long,
        x: Int,
        y: Int,
        z: Int,
        type: Material?,
        data: BlockData?,
        regionType: Int
    ): Boolean {
        val time = System.currentTimeMillis()
        if (time - respawnTime > 86400000L) {
            return true
        }
        return if (respawnTime < time && regionType == 0) {
            type != Material.SAND && type != Material.GRAVEL || this.plugin.server.getWorld(worldName).getBlockAt(
                x,
                y - 1,
                z
            ).type != Material.AIR
        } else respawnTime < time && regionType == 1 && this.plugin.server.getWorld(worldName).getBlockAt(
            x,
            y - 1,
            z
        ).type != Material.AIR
    }

    fun broadcastMessage(
        regionAlarmRadius: Int,
        regionAlarmMessage: String,
        worldName: String,
        x: Int,
        y: Int,
        z: Int
    ) {
        val location = Location(this.plugin.server.getWorld(worldName), x.toDouble(), y.toDouble(), z.toDouble())
        val onlinePlayers: Array<Player> = plugin.server.onlinePlayers.toTypedArray()
        val length = onlinePlayers.size
        var i = 0
        while (i < length) {
            val player = onlinePlayers[i]
            if (player.world.name.equals(
                    worldName,
                    ignoreCase = true
                ) && player.location.distance(location) <= regionAlarmRadius
            ) {
                this.plugin.log!!.sendPlayerWarn(player, regionAlarmMessage)
            }
            ++i
        }
    }
}
